Absorb prune into ensure

ibrasho opened the pull request:
### What does this do / why do we need it?

This PR updates `gps.WriteDepTree` and `Write` on `dep.*SafeWriter`to accept a `gps.PruneOptions` and a `*log.Logger` to determine the pruning options when writing `vendor/` and to enable verbose pruning.

The manifest is updated to accept a new table, `prune`. This table defines pruning options accepted by dep. A constructor for `dep.Manifest` is added to set the default PruneOptions.

By default, dep will always prune nested vendor directories.

Users can define 3 additional rules:
 * `non-go`: Prune non-Go files (will keep LICENSE & COPYING files)
 * `go-tests`: Prune Go test files
 * `unused-packages`: Prune unused Go packages.

### What should your reviewer look out for in this PR?

Design comments and correctness as usual. 

### Which issue(s) does this PR fix?

fixes #944

ibrasho commented:
@sdboyer I'm wondering how I should pass the `PruneOptions` to the `SafeWriter`?

I'm contemplating the choice between passing it in `NewSafeWriter` or when calling `SafeWriter.Write`... What do you think?

sdboyer commented:
> I'm contemplating the choice between passing it in `NewSafeWriter` or when calling `SafeWriter.Write`... What do you think?

eh, i don't have a strong feeling either way 🤷‍♂️

carolynvs commented on cmd/dep/gopath_scanner.go:
```diff
@@ -53,8 +53,9 @@ func (g *gopathScanner) InitializeRootManifestAndLock(rootM *dep.Manifest, rootL
-		Ovr:         make(gps.ProjectConstraints),
+		Constraints:  g.pd.constraints,
+		Ovr:          make(gps.ProjectConstraints),
+		PruneOptions: gps.PruneNestedVendorDirs,
```
Now that we want to have a default value for manifests (pruning nested vendor dirs), I'd prefer that we put that logic into a constructur `dep.NewManifest`, instead of relying upon every caller to remember to set it and get the default right.

Also, I didn't see that you updated the importers (they all make new manifests) and it's not clear to me what will even happen in that case if they leave the prune option off?

ibrasho replied:
I was basing this on the assumption that pruning is opt-in. So we only remove nested vendor directories by default.

carolynvs replied:
Sorry I'm late to the prune party and need a refresher course. When you say "pruning is opt-in" and also "we remove nested directories by default", I am not sure what will happen if someone doesn't explicitly set `PruneOptions`.

Basically I am concerned that if it's supposed to be a default, then it would be helpful to have it set by default in a constructor, rather than having to remember to set it everywhere as it's currently done in this PR. Some places aren't setting any `PruneOptions` (like the importers) and I can't really tell if they should, or if it doesn't matter because it's the default anyway.

If it's opt-in, do you mean that nothing happens (including removing nested vendor dirs) unless the user specifies a flag on the CLI or in the manifest? If so then why is it being set in code, that seems like a default, not opt-in...

None of this is criticism. I'm just confused. 😊

ibrasho replied:
I've already added a Manifest constructor based on your comments, seems like a great idea. 
I've updated the importers to use the new constructor. :grin:

I'll rebase and restructure the commits to allow sane review since this PR contains a lot of changes. 😓

carolynvs commented on internal/gps/prune.go:
```diff
@@ -0,0 +1,303 @@
+		}
+	}
+
+	// if (options & PruneNonGoFiles) != 0 {
```
Is this commented out because the PR is still a WIP?

ibrasho replied:
Yes. 😓 

I actually completed the PR but did a `git checkout -- .` by mistake and now I have to do it again. 😫 

I'm hoping that I'll be able to complete the remaining changes, again, tomorrow.

carolynvs replied:
Oh noes! Are any of your changes still hiding in the reflog perhaps?

carolynvs commented on manifest.go:
```diff
@@ -171,10 +196,11 @@ func readManifest(r io.Reader) (*Manifest, []error, error) {
+		Ovr:          make(gps.ProjectConstraints, len(raw.Overrides)),
+		Ignored:      raw.Ignored,
+		Required:     raw.Required,
+		PruneOptions: gps.PruneNestedVendorDirs,
```
Seems like it's not just a default but a required "base flag" that is always set and manifest cannot unset this flag?

ibrasho replied:
Good catch. I could change this so that coming through `dep/cmd/dep` always toggle that flag in `PruneOptions` when passing it to `SafeWriter.Write` or `gps.WriteDepTree`.

I understood from @sdboyer that we want `gps` to have control over this flag, but `dep` should always keep it on.

carolynvs replied:
ooooooh, knowing that gps may want to use it differently than dep makes this make a whole lot more sense! 

I don't know what's the best way to make the change, you are much closer to this than I am. But at a high level, having a way for dep to make sure that optional is always passed in properly seems useful.

carolynvs replied:
Don't let me tell you how to implement, I'm just asking questions and thinking out loud. 😃

ibrasho commented:
@sdboyer Just to clarify and confirm the effect of each pruning option:

`PruneNestedVendorOptions`: Remove any `vendor` directory existing at any level. This is always on.
`PruneNonGoFiles`: Remove any file that doesn't end in `.go`, except for files that start with either `LICENSE` or `COPYING`. (just in case files have an extension, e.g. `LICENSE.md`...).
`PruneGoTestFiles`: Remove any files that end in `_test.go`.
`PruneUnusedPackages`: This is the annoying one.  We can't delete an unused package if it contains an imported subpackage, so we do DFS and for each unimported Go package:
  * if more or more sub packages are imported, delete all files but keep the folders of the imported packages.
  * if no sub packages are imported, delete the directory and its content.

Finally, we always walk over `baseDir` and delete any empty directories. Is this necessary?

carolynvs commented on internal/gps/result.go:
```diff
@@ -42,37 +45,61 @@ type solution struct {
-func WriteDepTree(basedir string, l Lock, sm SourceManager, sv bool) error {
+// It requires a SourceManager to do the work, and takes a PruneOptions
+// indicating the pruning options required for the exported dependencies.
+func WriteDepTree(baseDir string, l Lock, sm SourceManager, prune PruneOptions, logger *log.Logger) error {
```
Oops, this looks like unmerged commits from #968 snuck in? Or did you intend to combine these PRs?

ibrasho replied:
This change is needed in both PRs since we need to log what was pruned.

carolynvs commented on cmd/dep/godep_importer.go:
```diff
@@ -88,9 +88,7 @@ func (g *godepImporter) load(projectDir string) error {
-	manifest := &dep.Manifest{
-		Constraints: make(gps.ProjectConstraints),
-	}
+	manifest := dep.NewManifest()
```
Yay! Thank you for the constructor. 😃

ibrasho commented:
Sorry for the slow updates. I was in quite the bind at work last week. 😩

Rebased and updated the commits. I'll be working on the tests now. 
@carolynvs & @sdboyer: Let me know if you have any further comments.

ibrasho commented:
The last commit is the result of running `dep ensure` with the prune options in `Gopkg.toml`.

ebati commented on internal/gps/result.go:
```diff
@@ -42,37 +45,61 @@ type solution struct {
+
+	if len(errCh) > 0 {
+		logger.Println("Failed to write dep tree. The following errors occurred:")
+		for err := range errCh {
```
Since channel is not closed, this will be blocked? Channel should be closed after `wg.Wait()` and return below should return an error since closed channel yields nil.

ibrasho replied:
Good catch. :+1:

ibrasho commented:
Some questions I've:
* All pruning operations are split into 2 parts: calculating files (or directories) and deleting. Is this necessary? Or can they be merged?
* Should I add a new list, I'm thinking `preserved-files`, to allow the user to specify files that should be kept regardless on the prune options? (values should be file paths relative to `vendor/`)

sdboyer commented on internal/gps/prune.go:
```diff
@@ -0,0 +1,367 @@
+)
+
+var (
+	// licenseFilePrefixes is a list of name prefixes for licesnse files.
```
typo: licesnse

sdboyer commented on internal/gps/result.go:
```diff
@@ -43,38 +45,65 @@ type solution struct {
 	}
 
-	// TODO(sdboyer) parallelize
+	var wg sync.WaitGroup
```
also doing the parallelizing here? in general, would prefer to keep down the number of major changes in a single PR.

ibrasho commented:
Closing this in favor of smaller PRs. :grin:
